l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,478 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import {EfficientCall} from "@matterlabs/zksync-contracts/l2/system-contracts/libraries/EfficientCall.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @title ConsensusRegistry
+/// @dev Manages consensus nodes and committees for the L2 consensus protocol,
+/// owned by a governance protocol. Nodes act as both validators and attesters,
+/// each playing a distinct role in the consensus process. This contract facilitates
+/// the rotation of validator and attester committees, which represent a subset of nodes
+/// expected to actively participate in the consensus process during a specific time window.
+contract ConsensusRegistry is Ownable2Step {
+    // An array to keep track of node owners.
+    address[] public nodeOwners;
+    // A map of node owners => nodes.
+    mapping(address => Node) public nodes;
+
+    uint256 validatorsCommit;
+    uint256 attestersCommit;
+    PendingNodeRemoval[] pendingRemovals;
+
+    struct Node {
+        AttesterAttr attesterLatest;
+        AttesterAttr attesterSnapshot;
+        uint256 attesterLastUpdateCommit;
+        ValidatorAttr validatorLatest;
+        ValidatorAttr validatorSnapshot;
+        uint256 validatorLastUpdateCommit;
+    }
+
+    /// @dev Represents a consensus node.
+    struct AttesterAttr {
+        // A flag stating if the node is active.
+        // Inactive attesters are not considered when selecting committees.
+        bool active;
+        // Attester's Voting weight.
+        uint32 weight;
+        // Attester's Secp256k1 public key.
+        Secp256k1PublicKey pubKey;
+    }
+
+    /// @dev Represents a consensus node.
+    struct ValidatorAttr {
+        // A flag stating if the node is active.
+        // Inactive validators are not considered when selecting committees.
+        bool active;
+        // Validator's voting weight.
+        uint32 weight;
+        // Validator's BLS12-381 public key.
+        BLS12_381PublicKey pubKey;
+        // Validator's Proof-of-possession (a signature over the public key).
+        BLS12_381Signature pop;
+    }
+
+    struct PendingNodeRemoval {
+        address nodeOwner;
+        bool pendingAttestersCommit;
+        bool pendingValidatorsCommit;
```

Comment:
A comment to show what these mean would be nice.

Reply:
Outdated.